Plugins: Improve instrumentation by adding metrics and tracing

xnyo opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

- [x] Adds tracing to the plugins backend GRPC client
- ~~[x] Passes an `X-Trace-Id` header to the plugin GRPC server~~
  - ~~Currently, this can't be used in any meaningful way for plugin developers. We may need to extend the SDK. Figuring out what to do with this.~~ -> Different scope, moved to https://github.com/grafana/grafana-plugin-sdk-go/issues/603
- [x] Add intrumentation to the plugin updater:
  - [x] Number of requests
  - [x] Number of failures
  - [x] Number of in-flight requests
  - [x] Latency histogram
- [x] Add tracing to the plugin updater
- [x] Add intrumentation to the grafana update checker
- [x] Add tracing to the grafana update checker 

**Why do we need this feature?**

Adds instrumentation and tracing to plugins and other plugin related features (like the update checker), 

**Who is this feature for?**

Plugin developers, Grafana operators

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #55464

**Special notes for your reviewer**:

xnyo committed:
WIP: Plugins tracing

xnyo committed:
Trace ID middleware

xnyo committed:
Add prometheus metrics and tracing to plugins updater

xnyo committed:
Add TODOs

xnyo committed:
Add instrumented http client

xnyo committed:
Add tracing to grafana update checker

marefr commented on pkg/services/pluginsintegration/clientmiddleware/tracing_header_middleware.go:
```diff
@@ -42,6 +46,22 @@ func (m *TracingHeaderMiddleware) applyHeaders(ctx context.Context, req backend.
 		}
 		req.SetHTTPHeader(headerName, gotVal)
 	}
+	m.applyTraceIDHeader(ctx, req)
```
I'm pretty sure the tracing identifier is already sent in gRPC metadata via https://github.com/grafana/grafana/blob/4def287e62e7d51bfbc62e062d580aa575fc60cd/pkg/plugins/backendplugin/grpcplugin/client.go#L42-L51 so these changes shouldn't be needed.

marefr commented on pkg/plugins/backendplugin/instrumentation/instrumentation.go:
```diff
@@ -30,8 +30,16 @@ var (
+	tracer tracing.Tracer, pluginCtx *backend.PluginContext,
+	endpoint string, fn func() error,
+) error {
+	ctx, span := tracer.Start(ctx, endpoint)
+	defer span.End()
```
I would suggest moving this to a tracing client middleware to clean up this function and dependencies to the plugin client

marefr reviewed (commented):
Happened to have a quick look at this. Some early feedback. We can discuss it later today if you want.

xnyo committed:
Merge branch 'main' into giuseppe/55464-plugins-instrumentation-improvements

xnyo committed:
Goimports

xnyo committed:
Moved plugins tracing to middleware

xnyo committed:
goimports, fix tests

xnyo committed:
Removed X-Trace-Id header

xnyo committed:
Fix comment in NewTracingHeaderMiddleware

xnyo committed:
Add metrics to instrumented http client

xnyo committed:
Add instrumented http client options

xnyo committed:
Merge branch 'main' into giuseppe/55464-plugins-instrumentation-improvements

xnyo committed:
Removed unused function

xnyo committed:
Switch to contextual logger

xnyo committed:
Refactoring, fix tests

xnyo committed:
Moved InstrumentedHTTPClient and PrometheusMetrics to their own package

xnyo committed:
Tracing middleware: handle errors

xnyo committed:
Report span status codes when recording errors

xnyo committed:
Merge branch 'main' into giuseppe/55464-plugins-instrumentation-improvements

xnyo committed:
Add tests for tracing middleware

xnyo committed:
Moved fakeSpan and fakeTracer to pkg/infra/tracing

xnyo committed:
Add TestHTTPClientTracing

xnyo committed:
Lint

wbrowne commented on pkg/services/pluginsintegration/clientmiddleware/tracing_middleware.go:
```diff
@@ -0,0 +1,86 @@
+func (m *TracingMiddleware) traceWrap(ctx context.Context, opName string) (context.Context, func(error)) {
+	ctx, span := m.tracer.Start(ctx, opName)
+	return ctx, func(err error) {
+		span.End()
```
This should probably be deffered since you're calling `SetStatus` and `RecordError` after it's ended, no? Or else just added after the `if` perhaps

wbrowne commented on pkg/services/updatechecker/instrumentation/httpclient.go:
```diff
@@ -0,0 +1,128 @@
+	"github.com/grafana/grafana/pkg/infra/tracing"
+)
+
+// InstrumentedHTTPClient is an HTTP client that wraps every request a span, tracking the method, url,
```
```suggestion
// InstrumentedHTTPClient is an HTTP client that wraps every request with a span, tracking the method, url,
```

wbrowne commented on pkg/services/updatechecker/grafana.go:
```diff
@@ -57,20 +72,36 @@ func (s *GrafanaService) Run(ctx context.Context) error {
 
-func (s *GrafanaService) checkForUpdates() {
-	resp, err := s.httpClient.Get("https://raw.githubusercontent.com/grafana/grafana/main/latest.json")
+func (s *GrafanaService) checkForUpdates(ctx context.Context) {
```
I think it may be cleaner to just duplicate code like so, instead of worrying about the `err` in the defer (shadowing, etc.)

```suggestion
func (s *GrafanaService) checkForUpdates(ctx context.Context) {
	var err error
	ctx, span := s.tracer.Start(ctx, "updatechecker.GrafanaService.checkForUpdates")
	defer span.End()

	ctxLogger := s.log.FromContext(ctx)
	ctxLogger.Debug("Checking for updates")
	resp, err := s.httpClient.Get(ctx, "https://raw.githubusercontent.com/grafana/grafana/main/latest.json")
	if err != nil {
		span.SetStatus(codes.Error, "update check failed: "+err.Error())
		span.RecordError(err)
		ctxLogger.Debug("Failed to get latest.json repo from github.com", "error", err)
		return
	}
	defer func() {
		if err := resp.Body.Close(); err != nil {
			ctxLogger.Warn("Failed to close response body", "err", err)
		}
	}()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		span.SetStatus(codes.Error, "update check failed: "+err.Error())
		span.RecordError(err)
		ctxLogger.Debug("Update check failed, reading response from github.com", "error", err)
		return
	}

	type latestJSON struct {
		Stable  string `json:"stable"`
		Testing string `json:"testing"`
	}
	var latest latestJSON
	err = json.Unmarshal(body, &latest)
	if err != nil {
		span.SetStatus(codes.Error, "update check failed: "+err.Error())
		span.RecordError(err)
		ctxLogger.Debug("Failed to unmarshal latest.json", "error", err)
		return
	}

	ctxLogger.Debug("Update check succeeded")

	s.mutex.Lock()
	defer s.mutex.Unlock()
	if strings.Contains(s.grafanaVersion, "-") {
		s.latestVersion = latest.Testing
		s.hasUpdate = !strings.HasPrefix(s.grafanaVersion, latest.Testing)
	} else {
		s.latestVersion = latest.Stable
		s.hasUpdate = latest.Stable != s.grafanaVersion
	}

	currVersion, err1 := version.NewVersion(s.grafanaVersion)
	latestVersion, err2 := version.NewVersion(s.latestVersion)
	if err1 == nil && err2 == nil {
		s.hasUpdate = currVersion.LessThan(latestVersion)
	}
}
```
```

wbrowne reviewed (commented):
Awesome stuff 🚀  - just caught a couple initial things worth mentioning

xnyo committed:
Changes after PR review

marefr commented on pkg/services/pluginsintegration/clientmiddleware/tracing_middleware_test.go:
```diff
@@ -0,0 +1,162 @@
+				return err
+			},
+		},
+		// TODO: RunStream?
```
Yes please, if possible

marefr commented on pkg/services/updatechecker/grafana.go:
```diff
@@ -57,20 +72,36 @@ func (s *GrafanaService) Run(ctx context.Context) error {
+	ctxLogger := s.log.FromContext(ctx)
+	defer func() {
+		if err != nil {
+			span.SetStatus(codes.Error, "update check failed: "+err.Error())
```
Nit. fmt.Sprintf instead?

marefr commented on pkg/services/updatechecker/grafana.go:
```diff
@@ -57,20 +72,36 @@ func (s *GrafanaService) Run(ctx context.Context) error {
+		if err != nil {
+			span.SetStatus(codes.Error, "update check failed: "+err.Error())
+			span.RecordError(err)
+			ctxLogger.Debug("Update check failed")
```
```suggestion
			ctxLogger.Error("Update check failed", "error", err)
```

marefr commented on pkg/services/updatechecker/grafana.go:
```diff
@@ -57,20 +72,36 @@ func (s *GrafanaService) Run(ctx context.Context) error {
+	resp, err := s.httpClient.Get(ctx, "https://raw.githubusercontent.com/grafana/grafana/main/latest.json")
 	if err != nil {
-		s.log.Debug("Failed to get latest.json repo from github.com", "error", err)
+		ctxLogger.Debug("Failed to get latest.json repo from github.com", "error", err)
```
Error level? Same for some additional errors below.

marefr commented on pkg/services/updatechecker/grafana.go:
```diff
@@ -40,15 +55,15 @@ func (s *GrafanaService) IsDisabled() bool {
 		select {
 		case <-ticker.C:
-			s.checkForUpdates()
+			s.checkForUpdates(ctx)
```
Code might be simplified by returning an error here, checking it and log? Same applies for plugins update checker

xnyo replied:
Nice catch, I have updated the code :+1:

marefr commented on pkg/services/updatechecker/instrumentation/httpclient.go:
```diff
@@ -0,0 +1,128 @@
+
+// InstrumentedHTTPClient is an HTTP client that wraps every request with a span, tracking the method, url,
+// and response code as span attributes.
+type InstrumentedHTTPClient struct {
```
This definitely works. But would consider reusing the SDK httpclient library we already have that supports middlewares. Using [httpclient.New](https://github.com/grafana/grafana-plugin-sdk-go/blob/4344c93ea2747c85a2d3d18635be0766e56af4be/backend/httpclient/http_client.go#L18). Then maybe this [tracing middleware](https://github.com/grafana/grafana/blob/f0333ac41f0d2b0c0107f5794b7d5c7ca8aa0401/pkg/infra/httpclient/httpclientprovider/tracing_middleware.go#L21) can be resued?

WDYT?

xnyo replied:
Switched to httpclient tracing middleware and a new prometheus metrics middleware as suggested :)

marefr commented on pkg/services/updatechecker/instrumentation/httpclient.go:
```diff
@@ -0,0 +1,128 @@
+	inFlightGauge prometheus.Gauge,
+) InstrumentedHTTPClientOption {
+	return func(cl *InstrumentedHTTPClient) {
+		cl.metrics = &PrometheusMetrics{
```
Maybe personal opinion. Given we have logs capturing this don't think we need metrics for it. Might simplify the code a bit as well.

WDYT?

xnyo replied:
Removed as suggested because we can create those metrics from loki as well :+1:

marefr reviewed (commented):
Great work. Added some comments/suggestions, mostly minor things. 

For the update/plugins checker would like to see logs including the duration it took to run since these can be used as metrics with Loki. Example code:
https://github.com/grafana/grafana/blob/f0333ac41f0d2b0c0107f5794b7d5c7ca8aa0401/pkg/services/cleanup/cleanup.go#L108-L120

xnyo committed:
Tests: Made "ended" in FakeSpan private, allow calling End only once

xnyo committed:
Testing: panic in FakeSpan if span already ended

xnyo committed:
Refactoring: Simplify Grafana updater checks

xnyo committed:
Refactoring: Simplify plugins updater error checks and logs

xnyo committed:
Fix wrong call to checkForUpdates -> instrumentedCheckForUpdates

xnyo committed:
Tests: Fix wrong call to checkForUpdates -> instrumentedCheckForUpdates

xnyo committed:
Merge branch 'main' of github.com:grafana/grafana into giuseppe/55464-plugins-instrumentation-improvements

xnyo committed:
Log update checks duration, use Info log level for check succeeded logs

xnyo committed:
Add plugin context span attributes in tracing_middleware

xnyo committed:
Refactor prometheus metrics as httpclient middleware

xnyo committed:
Fix call to ProvidePluginsService in plugins_test.go

xnyo committed:
Propagate context to update checker outgoing http requests

xnyo committed:
Plugin client tracing middleware: Removed operation name in status

xnyo committed:
Merge branch 'main' into giuseppe/55464-plugins-instrumentation-improvements

xnyo committed:
Fix tests

xnyo committed:
Goimports tracing_middleware.go

xnyo committed:
Merge branch 'main' into giuseppe/55464-plugins-instrumentation-improvements

xnyo committed:
Goimports

xnyo committed:
Merge branch 'main' into giuseppe/55464-plugins-instrumentation-improvements

xnyo committed:
Fix imports

xnyo committed:
Merge branch 'main' into giuseppe/55464-plugins-instrumentation-improvements

xnyo committed:
Changed span name to plugins client middleware

xnyo committed:
Add span name assertion in TestTracingMiddleware

xnyo committed:
Merge branch 'main' into giuseppe/55464-plugins-instrumentation-improvements

xnyo commented on pkg/services/pluginsintegration/clientmiddleware/tracing_middleware.go:
```diff
@@ -0,0 +1,86 @@
+}
+
+func (m *TracingMiddleware) traceWrap(ctx context.Context, opName string) (context.Context, func(error)) {
+	ctx, span := m.tracer.Start(ctx, opName)
```
:+1:  Added `PluginClient.` as prefix

wbrowne reviewed (approved):
LGTM. Awesome work!

BTW is this just an oversight now? https://github.com/grafana/grafana/pull/61035/files#diff-8362b35f918aafddca8ace7834a44a1bf9a7b9ca187525a3e4459eae4589ab3aR86

xnyo committed:
Merge branch 'main' into giuseppe/55464-plugins-instrumentation-improvements

xnyo committed:
Removed Prometheus metrics middleware from grafana and plugins updatechecker

xnyo committed:
Add span attributes for ds name, type, uid, panel and dashboard ids

xnyo committed:
Fix http header reading in tracing middlewares

svennergr commented on pkg/services/pluginsintegration/clientmiddleware/tracing_middleware.go:
```diff
@@ -29,17 +31,42 @@ type TracingMiddleware struct {
+			attribute string
+		}{
+			{"X-Panel-Id", "panel_id"},
+			{"X-Dashboard-Id", "dashboard_id"},
```
Just a quick note: `X-Dashboard-Id` is a legacy header and was changed to `X-Dashboard-Uid`, which is not an `int` but a `string`, a while ago.

Sorry for linking to the headers from ds_proxy previously, I think those are the relevant ones: https://github.com/grafana/grafana/blob/b1a7e1d33d6bc04432849be251c80deec984dd54/pkg/services/query/query.go#L26-L32

xnyo committed:
Use contexthandler.FromContext, add X-Query-Group-Id

xnyo committed:
Add test for RunStream

xnyo commented on pkg/services/pluginsintegration/clientmiddleware/tracing_middleware.go:
```diff
@@ -0,0 +1,133 @@
+	// Start span
+	ctx, span := m.tracer.Start(ctx, "PluginClient."+opName)
+
+	// Attach some plugin context information to span
+	span.SetAttributes("plugin_id", pluginContext.PluginID, attribute.String("plugin_id", pluginContext.PluginID))
+	span.SetAttributes("org_id", pluginContext.OrgID, attribute.Int64("org_id", pluginContext.OrgID))
+	if settings := pluginContext.DataSourceInstanceSettings; settings != nil {
+		span.SetAttributes("datasource_name", settings.Name, attribute.Key("datasource_name").String(settings.Name))
+		span.SetAttributes("datasource_type", settings.Type, attribute.Key("datasource_type").String(settings.Type))
+		span.SetAttributes("datasource_uid", settings.UID, attribute.Key("datasource_uid").String(settings.UID))
+	}
+	if u := pluginContext.User; u != nil {
+		span.SetAttributes("user", u.Login, attribute.String("user", u.Login))
+	}
+
+	// Additional attributes from http headers
+	if reqCtx := contexthandler.FromContext(ctx); reqCtx != nil && len(reqCtx.Req.Header) > 0 {
+		if v, err := strconv.Atoi(reqCtx.Req.Header.Get(query.HeaderPanelID)); err == nil {
+			span.SetAttributes("panel_id", v, attribute.Key("panel_id").Int(v))
+		}
+		setSpanAttributeFromHTTPHeader(reqCtx, span, "query_group_id", query.HeaderQueryGroupID)
+		setSpanAttributeFromHTTPHeader(reqCtx, span, "dashboard_uid", query.HeaderDashboardUID)
+	}
```
@svennergr Please would you mind double-checking those are the attributes you are interested in or if I have missed anything? Thanks :smiley:

xnyo commented:
@wbrowne @marefr please would you mind reviewing the changes again since your last review?

@wbrowne since your last review I have:
- Added tracing test case for RunStream method, which I forgot to add before
- Added more span attributes (datasource uid, dashboard uid and panel id...)

marefr commented on pkg/services/pluginsintegration/clientmiddleware/tracing_middleware.go:
```diff
@@ -0,0 +1,133 @@
+func (m *TracingMiddleware) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
+	ctx, end := m.traceWrap(ctx, req.PluginContext, "queryData")
+	resp, err := m.next.QueryData(ctx, req)
+	end(err)
```
Use defer in case m.next.QueryData panics? Same applies for all methods below.

xnyo replied:
Nice catch! I have changed the code and added tests for panics :+1:

marefr commented on pkg/services/pluginsintegration/clientmiddleware/tracing_middleware.go:
```diff
@@ -0,0 +1,133 @@
+	span.SetAttributes("org_id", pluginContext.OrgID, attribute.Int64("org_id", pluginContext.OrgID))
+	if settings := pluginContext.DataSourceInstanceSettings; settings != nil {
+		span.SetAttributes("datasource_name", settings.Name, attribute.Key("datasource_name").String(settings.Name))
+		span.SetAttributes("datasource_type", settings.Type, attribute.Key("datasource_type").String(settings.Type))
```
Nit. You can skip this since it's the same as plugin_id I'm pretty sure.

```suggestion
```

xnyo replied:
You're right. I have removed it :+1:

marefr commented on pkg/services/pluginsintegration/clientmiddleware/tracing_middleware.go:
```diff
@@ -0,0 +1,133 @@
+	}
+
+	// Additional attributes from http headers
+	if reqCtx := contexthandler.FromContext(ctx); reqCtx != nil && len(reqCtx.Req.Header) > 0 {
```
Nit. Check if `reqCtx.Req` not nil.

marefr reviewed (changes requested):
Great work. Some minor things left.

xnyo committed:
Merge branch 'main' into giuseppe/55464-plugins-instrumentation-improvements

xnyo committed:
Fix imports

xnyo committed:
Changes from PR review

xnyo committed:
TestTracingMiddleware: Changed assert to require for didPanic assertion

marefr reviewed (approved):
LGTM great work

xnyo committed:
Lint

xnyo committed:
Merge branch 'main' into giuseppe/55464-plugins-instrumentation-improvements

xnyo committed:
Fix imports

xnyo commented on pkg/infra/tracing/test_helper.go:
```diff
@@ -1,7 +1,86 @@
+func (t *FakeTracer) Inject(ctx context.Context, header http.Header, span Span) {
+}
+
+func NewFakeTracer() *FakeTracer {
```
:+1:  I will add in a follow-up PR
